### DISCUSSION TITLE
Installed module is not found by IronPython

### DISCUSSION THREAD

**Discussion by hesperaux** [IronLanguages/ironpython3#1646] [Q&A]
Hello,

I am using the code below to try to run a python script that connects to a VISA-interfaced voltmeter. The python script works fine in Python 3.7. But using ipy.exe or embedding it in a .net7 program, it is not working.

Here is the code:
```
        public static void Open(object caller, string instrumentId)
        {
            ScriptEngine engine = Python.CreateEngine();
            Dictionary<object, ScriptScope> _scriptScopes = new();
            ICollection<string>? paths = engine.GetSearchPaths();
            paths.Add(@"C:\Program Files\IronPython 3.4\Lib\site-packages");
            paths.Add(@"C:\Program Files\IronPython 3.4\Lib\");
            engine.SetSearchPaths(paths);
            engine.ImportModule("pyvisa_py");
            _scriptScopes.Add(caller, engine.CreateScope());

            engine.ExecuteFile(@"pyvisa-wrap.py", _scriptScopes[caller]);
        }
```

And the python script (simplified):
```
import pyvisa
import pyvisa_py

rm = pyvisa.ResourceManager('@py')
print(rm.list_resources())
```

The result is an ImportException:
`IronPython.Runtime.Exceptions.ImportException: no module named pyvisa_py`

I have installed pyvisa, pyusb, and pyvisa-py packages into the global scope using an admin powershell with ipy.exe:
```
C:\Program Files\IronPython 3.4> ./ipy.exe -m pip install pyusb==1.1.1
C:\Program Files\IronPython 3.4> ./ipy.exe -m pip install pyvisa==1.10.1
C:\Program Files\IronPython 3.4> ./ipy.exe -m pip install pyvisa-py==0.4.1
```

If I run the script with ipy.exe, I get the same problem:
```
C:\Program Files\IronPython 3.4> .\ipy.exe C:\Repos\sw\py\pyvisa-wrapper\main.py
Traceback (most recent call last):
  File "C:\Repos\sw\py\pyvisa-wrapper\main.py", line 5, in <module>
ImportError: No module named 'pyvisa_py'
```

The contents of my site-packages:
![image](https://user-images.githubusercontent.com/1521481/211908914-3114562b-e06a-46b8-8a73-a4b76404a076.png)

I would be grateful for any guidance. Thank you!

**Comment by slozier** [comment]
I think you can just drop the `import pyvisa_py`? Not seeing that as being required in any of the docs.

The `pyvisa-py` appears to be named `pyvisa-py` in version 0.4.1 and `pyvisa_py` in version 0.6.0 (which I assume is what you're using with Python 3.7).

**Reply by hesperaux** [reply]
I did try this, and it will run this way, however, it doesn't actually work. It behaves as if pyvisa-py was never loaded (the visalib.get_library_paths() function returns 'unset' instead of 'py'). Therefore, I put this in here explicitly to see if it was working, and that's when I found out that it won't import. So I was working on the assumption that this was happening within the PyVisa library and was just being gracefully handled, since that library attempts to search for supported backends cleanly.

Apparently the only way to import pyvisa-py is by importing "pyvisa_py". Python syntax does not allow hyphens in the code for the package name, it seems?

I have tried to import pyvisa-py, though, and that doesn't "compile." (see screenshot)
![image](https://user-images.githubusercontent.com/1521481/212221678-54d68181-afe0-4e1b-844f-b3ae6e3a70a6.png)

All that being said, with the import statement removed, it could be that another issue is preventing it from working. pyvisa-py uses pyusb, which in turn uses libusbX.dll, where X is 0 or 1. This DLL (or SO if on linux) needs to be in the libpath for the interpreter for it to work. Is there any special magic required to add C/C++ DLLs to the lib path for iron python when used as a NuGet package (or using ipy.exe for that matter)?

On my vanilla python virtual env, I'm using pyvisa-py 0.5.3 by the way, not 0.6.0. And it is with python version 3.8.0.

I've also tried to copy the package files into a local directory and add those, rather than the path in program files. That doesn't help either. This is actually what lead me to try the global site-packages location (which, requiring admin on windows, would've been less desirable).

**Reply by slozier** [reply]
If you think it might help, you could get the module to load using importlib:
```py
import importlib
pyvisa_py = importlib.import_module("pyvisa-py")
```
but I don't really think that's the issue. Just installing the pyvisa-py module seems to be enough to make it show up in the pyvisa module (I can see it by running this on the command line `ipy -m visa info`). It could very well by a pyusb issue, but I have no idea how I would go about helping you with that...

**Reply by hesperaux** [reply]
Thanks for the help.
Using importlib, it does import successfully. Unfortunately, that doesn't seem to be the issue.
Since this is outside the scope of ironpython as of right now, I'll investigate how to convince PyUSB to work properly in this environment. I'm guessing it is not discovering the libusb library correctly.

If I discover anything useful, I'll add it to this discussion in case anyone else comes around looking for it. I may end up having to pivot to a different design. If only vendors would update their libraries to support modern .NET, instead of .NET Framework 3.5....

**Comment by hesperaux** [comment]
Just a little update...the python solution was too tenuous so we eventually discovered that Rohde & Schwarz has a .NET nuget package. Overlooked this initially cuz I thought it was python only.
For anyone else that ends up at this page, it runs fine in .NET 7:
https://rsinstrumentcsharp.readthedocs.io/en/latest/StepByStepGuide.html

Edit: (It also runs on Raspberry Pi, which was something we were hoping for)
